docs/proposals/build-hooks.md

```diff
@@ -0,0 +1,218 @@
+# Build Hooks
+
+## Abstract
+
+A proposal for supporting the execution of hooks in the build lifecycle.
+
+
+## Motivation
+
+Users need to execute custom actions at different stages of the build process.
+OpenShift provides full flexibility on how builds are performed through the
+*Custom* build strategy, and *Source-To-Image* builds also allow customization
+via its `assemble` script. However, as we'll see, those extensibility points
+have their limitations.
+
+At the time of writing, **Deployments** have a hook mechanism that supports the
+execution of arbitrary commands before and after a deployment occurs,
+individually supported by each *Deployment Strategy*, namely *Recreate* and
+*Rolling*.
+The third existing strategy, *Custom*, does not support hooks.
+The implementation is done in the strategy level as to allow different
+strategies to support a potentially different set of hooks.
+
+Goals of this design:
+
+1. Identify build hook use cases
+2. Discuss and describe implementation details
+
+
+## Use cases
+
+### 1. Build pipeline
+
+As an application developer, I want to be able to [trigger a build after an
+upstream build completes](https://github.com/openshift/origin/issues/1228#issue-59933179),
+creating a build pipeline.
+
+
+### 2. Integration with external system
+
+As an application developer, I want to be notified or notify other people after
+a build completes, email or otherwise, including the build result (success or
+failure).
```

Comment:
If you need to report it to a webhook.  If you trigger something based on a
build completing, not being able to exactly use the output of the build is
unacceptable (ie it's a bug if you can't get the correct image).

On Jul 16, 2015, at 5:20 PM, Rodolfo Carvalho <notifications@github.com>
wrote:

In docs/proposals/build-hooks.md
<https://github.com/openshift/origin/pull/3736#discussion_r34839212>:

> +
> +
> +## Use cases
> +
> +### 1. Build pipeline
> +
> +As an application developer, I want to be able to [trigger a build after an
> +upstream build completes](https://github.com/openshift/origin/issues/1228#issue-59933179),
> +creating a build pipeline.
> +
> +
> +### 2. Integration with external system
> +
> +As an application developer, I want to be notified or notify other people after
> +a build completes, email or otherwise, including the build result (success or
> +failure).

BTW -- do we have any real use case for consuming the image ID, or is it
just something we believe might be useful?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3736/files#r34839212>.